src/EventStore.Core/TransactionLog/Scavenging/Throttle.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using EventStore.Common.Log;
+
+namespace EventStore.Core.TransactionLog.Scavenging {
+	// Call Rest from time to time and this will rest a suitable amount of time
+	// to achieve an overall % time spent working approximately equal to activePercent.
+	public class Throttle {
+		protected static readonly ILogger Log = LogManager.GetLoggerFor<Throttle>();
+		private readonly Stopwatch _stopwatch;
+		private readonly double _minimumRestMs;
+		private readonly double _logThresholdMs;
+		private readonly double _restFactor;
+		private readonly ManualResetEventSlim _mres;
+
+		private double _totalRestingTimeMs;
+
+		public Throttle(
+			TimeSpan minimumRest,
+			TimeSpan restLoggingThreshold,
+			double activePercent) {
+
+			if (activePercent <= 0 || 100 < activePercent)
+				throw new ArgumentOutOfRangeException(nameof(activePercent), activePercent, null);
+
+			_stopwatch = new Stopwatch();
+			_minimumRestMs = minimumRest.TotalMilliseconds;
+			_logThresholdMs = restLoggingThreshold.TotalMilliseconds;
+			_restFactor = 1 - (activePercent / 100);
+
+			_mres = new ManualResetEventSlim();
```

Comment:
do we want to set ManualResetEventSlim's spinCount to 0 to ensure we're immediately releasing the thread?

Reply:
done :+1:

Reply:
thanks!